### DISCUSSION TITLE
compatibility with pkgload::load_all()

### DISCUSSION THREAD

**Discussion by wlandau** [ropensci/targets#1211] [Ideas]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

@krlmlr, we have been emailing about compatibility between `targets` and `pkgload::load_all()`. I posted this discussion to make it easier to share code.

Let's say we have a package with this `DESCRIPTION` file:

```
Package: test
Type: Package
Title: What the Package Does (Title Case)
Version: 0.1.0
Author: Who wrote it
Maintainer: The package maintainer <yourself@somewhere.net>
Description: More about what it does (maybe more than one line)
    Use four spaces when indenting paragraphs within the Description.
License: What license is it under?
Encoding: UTF-8
LazyData: true
RoxygenNote: 7.3.0
```

and this `R/functions.R` file:

```r
f <- function() {
  Sys.getpid()
}
```

If we were to pursue integration with `load_all()`, I believe we should use a wrapper function around existing functionality for installed packages: https://books.ropensci.org/targets/packages.html#r-packages-as-projects. The wrapper I propose is `tar_load_all()` defined below, which sets the `envir`, `imports`, and `packages` options so that the pipeline uses the package namespace instead of the global environment. Here is the `_targets.R` file:

```r
library(targets)
tar_load_all <- function() {
  eval(parse(text = "pkgload::load_all()"))
  name <- pkgload::pkg_name()
  packages <- setdiff(tar_option_get("packages"), name)
  envir <- pkgload::pkg_ns()
  imports <- unique(c(tar_option_get("imports"), name))
  tar_option_set(packages = packages, envir = envir, imports = imports)
  invisible()
}
tar_load_all()
print(Sys.getpid())
list(
  tar_target(x, f())
)
```

When I inspect and run the pipeline locally, everything seems to work.

```r
library(targets)
tar_visnetwork() # Verify that the static code analysis detects f().
```

![Screenshot 2024-01-23 at 10 23 42 AM](https://github.com/ropensci/targets/assets/1580860/b68cb86c-eba8-48af-beae-2d347fd3456d)

```r
tar_make()
#> ℹ Loading test
#> [1] 36475
#> ▶ dispatched target x
#> ● completed target x [0 seconds]
#> ▶ completed pipeline [0.051 seconds]

tar_read(x)
#> [1] 36475
```

However, when the target runs in a [`crew` worker](https://books.ropensci.org/targets/crew.html), the functions in the namespace are missing.

```r
# _targets.R file with a {crew} controller:
library(targets)
tar_load_all <- function() {
  eval(parse(text = "pkgload::load_all()"))
  name <- pkgload::pkg_name()
  packages <- setdiff(tar_option_get("packages"), name)
  envir <- pkgload::pkg_ns()
  imports <- unique(c(tar_option_get("imports"), name))
  tar_option_set(packages = packages, envir = envir, imports = imports)
  invisible()
}
tar_load_all()

# Add a {crew} controller for parallel computing:
tar_option_set(controller = crew::crew_controller_local(workers = 1L))

print(Sys.getpid())
list(
  tar_target(x, f())
)
```

```r
# R console:
tar_destroy()
tar_make()
#> ℹ Loading test
#> [1] 36576
#> ▶ dispatched target x
#> ✖ errored target x
#> ✖ errored pipeline [1.638 seconds]
#> Error:
#> ! Error running targets::tar_make()
#> Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
#> Debugging guide: https://books.ropensci.org/targets/debugging.html
#> How to ask for help: https://books.ropensci.org/targets/help.html
#> Last error message:
#>     could not find function "f"
#> Last error traceback:
#>     .handleSimpleError(function (condition)  {     state$error <- build_mess...
#>     h(simpleError(msg, call))
```

The reason this happens is that `targets` is treating `tar_option_get("envir")` as a simple environment when it exports data to parallel workers:

https://github.com/ropensci/targets/blob/8ab431908ea584f990fc2727cd4e2aaabe3d0cb2/R/class_crew.R#L104-L167

In other words, the `envir` argument passed to `targets::target_run_worker()` is pretty much just `pkgload::pkg_ns()`, and somehow it becomes empty when sent to a worker.

I can isolate this behavior in a much simpler example using `callr` from the root of the package file system:

```r
pkgload::load_all()
callr::r(
  function(envir) {
    eval(f(), envir = envir)
  },
  args = list(envir = pkgload::pkg_ns())
)
#> Error: 
#> ! in callr subprocess.
#> Caused by error in `f()`:
#> ! could not find function "f"
#> Type .Last.error to see the more details.
```

If you know of a better way of marshaling package namespaces in a way that works with `pkgload::load_all()` and does not conflict with non-package environments, then maybe `tar_load_all()` could become part of `targets` for people who want `pkgload::load_all()`-like behavior. Otherwise, I am not sure how to make this work for a package that is not actually installed.

**Comment by krlmlr** [comment]
Nice summary, thanks! IIUC, remote workers never access the source code but need everything gift-wrapped so that they can just run it? That was new to me.

Let's find a way to make it work, then! `tar_load_all()` looks like a good compromise to me, but perhaps it's possible to just make `load_all()` work too, e.g., by searching for `.__DEVTOOLS__` in the package namespaces?

**Reply by wlandau** [reply]
> IIUC, remote workers never access the source code but need everything gift-wrapped so that they can just run it? 

Yes, exactly. For example, in the cloud pipelines I am running with `targets` and `crew.aws.batch`, the Docker containers of the workers do not have access to local files can cannot `load_all()` themselves.

> Let's find a way to make it work, then! tar_load_all() looks like a good compromise to me, but perhaps it's possible to just make load_all() work too, e.g., by searching for .__DEVTOOLS__ in the package namespaces?

`._DEVTOOLS__` could be useful for detecting whether `tar_option_set("envir")` was created by `pkgload::load_all()`. (Or maybe `.__NAMESPACE__` more generally if we want something that will work with both installed and uninstalled packages.) But as far as making `pkgload::load_all()` "just work"... even if it is possible, it would require runtime conditional logic to choose the defaults of many `tar_option_set()` options. In my mind, this is too much technical debt to have at a low level. (Related: unlike `drake`, static branching is not implemented in the package itself. The implementation is a bunch of wrapper functions in [`tarchetypes`](https://docs.ropensci.org/tarchetypes). I prefer to add these kinds of features using extra layers rather than deep infrastructure when possible.)

**Comment by asadow** [comment]
This is a very exciting idea. I frequently bring up a targets project for data exploration and wish I could `load_all()` instead of writing many library calls.

`tar_load_all()` however is a bit confusing with `tar_load()` already for loading targets.

**Comment by cpauvert** [comment]
Hi, 
thanks @wlandau for developing and continuously maintaining `{targets}`!

I'm still learning, therefore  I cannot offer any solutions regarding the topic at hand. However, it made me think of something. During package development, `pkgload::load_all()` is used regularly to test and evaluate the WIP functions, and when some of these functions are [targets factories](https://wlandau.github.io/targetopia/contributing.html#target-factories), examples cannot be ran as `targets` crash with:

```
# edited for clarity
! Error running targets::tar_make()
Last error message:
    there is no package called ‘wip_pkg_loaded_locally_but_not_installed’
```

I looked up the examples of Targetopia packages as well as dicussions, without a clue. I have a feeling that a wrapper like `tar_load_all` would enable this. 

Let me know if this is off-topic and would require its own thread (with more details, reprex etc), I do not want to pollute this discussion but merely showing interest and bringing another angle.

BEst,
